### DISCUSSION TITLE
WS client in node bindings?

### DISCUSSION THREAD

**Discussion by sarumjanuch** [uNetworking/uWebSockets.js#46] [Q&A] [answered]
Hi there, first of all, I want to thank you for all the effort you are putting into this, and for sharing all your work via open source. I am following all your work since early versions of original uws for node. I was really shocked when was benchmarking new uWebsockets.js against golang ws and http implementations, in a day of new version release. I wasn't able to believe in number I am getting in node.js. You are actually the person who woken up me, and now I am working in benchmark oriented development process, where i am testing all my node_modules crap before i am starting to use it. I was frustrated by how bad everything is in node world, and how blindly people just following stars count. Realizing the mess around, i quit front-end development and started to learn C :). So thank you very much again. I just wanted to ask is there WS client available in uWebSocket.js? In the original version, it was pretty straight forward, how to connect to ws using uws but now i am unable to find a way, I only see how to use the server. Sorry for such a question, but I am not C++ dev, and it is pretty hard for me to answer this question, digging in your code.

**Comment by ghost** [comment, marked as answer]
Thanks for the nice words, there's no client support yet but if I get the time there might come such support.

The reason this project focuses on server-only is because most companies who struggle with WebSocket performance struggle in the server department where many connections are held. Clients often only hold a handful, making less of an impact what library you use. You could use websockets/ws as client as long as you use µWS.js as server.

I think the entire point of WebSockets is to communicate browser <-> server so you already have the client in the browser.

**Comment by sarumjanuch** [comment]
Your library has very strong potential. I personally use it for node.js IPC (old one). First of all, WS provides a very thin layer on top of TCP\TLS - WS frames, this helps to have messages, and you don't need to create some StreamReaders, and parse stream into messages in node land. This is very convenient. Second, the performance is crazy. I made some very simple benchmarks more than half a year ago, where I made one process server, and second process client on the same machine (in case of ZMQ it was DEALER-ROUTER, and in case of NATS it was one publisher one subscriber connected to NATS server), and was measuring time to receive 5 000 000 tiny messages (I know, that there are many quirks in measuring such way, but this is just to get approximate order of magnitude):

https://github.com/uNetworking/uWebSockets-bindings/tree/master/nodejs
uWS self dep.
received 5 000 000 msgs in 3558.322ms
~ 1 400 000 msg/sec

TCP std lib.
received 5 000 000 msgs in 17092.544ms
~ 294 000 msg/sec

https://github.com/nats-io/node-nats
NATS self dep.
routed 5 000 000 msgs in 17343.417ms
~ 289 000 msg/sec

https://github.com/zeromq/zeromq.js/
ZeroMQ 63 dependencies.
received 5 000 000 msgs in 18144.784ms
~ 276 000 msg/sec

Unix Domain Sockets std lib
received 5 000 000 msgs in 18261.213.ms
~ 274 000 msg/sec

https://github.com/rolftimmermans/zeromq-ng
_**!!!! https://github.com/rolftimmermans/zeromq-ng/issues/32 - more details as I did not re-test after changes discussed in issue, so most likely this is not accurate.**_
ZeroMQ-NG 110 dependencies.
received 5 000 000 msgs in 35261ms
~ 142 000 msg/sec

I have noticed socket.io-enterprise in your repo, I believe this will have a lot of sense in node land if I understand it correctly.
